rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,379 @@
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- values
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- ip
+
+The instruction pointer points to the next instruction to evaluate.
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the stack.
+
+Since Vrl allows for named parameters and optional parameters, the compiler
+will need to ensure the bytecode evaluates each parameter in a specific order
+(likely the order declared in the function). Bytecode will need to be emmitted
+for parameters that are not specified in the Vrl script to add a default value
+to the stack.
+
+### Flow analysis
+
+One of the expensive operations is path lookup.
+
+There may be certain situations where this can be improved by analyising the
+flow of data at compile time.
+
+For example, with this code:
+
+```coffee
+.thing[3].thang = 3
+log(.thing[3].thang)
+```
+
+the compiler can work out that the variable written to and read from is the
+same and can thus store that data in an array, any access can be made via a
+quick O(1) lookup using the index to the array.
+
+The flow analysis can become fairly complex.
+
+```coffee
+.message = parse_json!(.message)
+.message[3] = parse_json!(.message[2])
+log(.message[3].thing)
+```
+
+Functions like `parse_json` would need access to this array and hints as to
+what paths should be stored.
+
+A lot more thought needs to go into this before we can consider implementing
+it.
+
+#### Use a Bump Allocator?
+
+We could use a library such as [bumpalo](https://crates.io/crates/bumpalo).
+
+A bump allocator will allocate memory up front. This memory will then be used
+during the runtime. At the end of each transform the bump pointer is reset to
+the start - effectively deallocating all the memory used during the transform
+with very little cost.
+
+The downside to a bump allocator is we need to make sure sufficient memory is
+allocated up front.
+
+This is probably not ready yet until this PR for `BTreeMap` lands.
+https://github.com/rust-lang/rust/pull/77438
+
+
+### Optimization
+
+With the code as a single dimension array of Bytecode, it could be possible to
+scan the code for patterns and reorganise the Bytecode so it can run in a more
+optimal way.
```

Comment:
I looked into doing this kind of pattern-based optimization briefly a while back and (with the caveat that I was _very_ new to VRL internals) I had a hard time finding a good place to do it. IIRC, there were a few more layers of indirection around the AST representation than would have been conducive to expressing those patterns in a simple way.

It's possible that I just didn't know the code well enough or that things have changed in the meantime, but even so it seems like it would be a valuable goal to have a simple, relatively tight representation of the AST available that makes experimenting with these kinds of optimizations as easy as possible.